File: src/types/prerequisites/PrerequisiteGroups.ts
import * as DB from "tsondb/schema/dsl"
import { ActivatablePrerequisite } from "./single/ActivatablePrerequisite.js"
import { AncestorBloodPrerequisite } from "./single/AncestorBloodPrerequisite.js"
import { AnimistPowerPrerequisite } from "./single/AnimistPowerPrerequisite.js"
import { CulturePrerequisite } from "./single/CulturePrerequisite.js"
import { EnhancementPrerequisite } from "./single/EnhancementPrerequisite.js"
import { InfluencePrerequisite } from "./single/InfluencePrerequisite.js"
import { PactPrerequisite } from "./single/PactPrerequisite.js"
import { PersonalityTraitPrerequisite } from "./single/PersonalityTraitPrerequisite.js"
import { PrimaryAttributePrerequisite } from "./single/PrimaryAttributePrerequisite.js"
import { PublicationPrerequisite } from "./single/PublicationPrerequisite.js"
import { RacePrerequisite } from "./single/RacePrerequisite.js"
import { RatedMinimumNumberPrerequisite } from "./single/RatedMinimumNumberPrerequisite.js"
import { RatedPrerequisite } from "./single/RatedPrerequisite.js"
import { RatedSumPrerequisite } from "./single/RatedSumPrerequisite.js"
import { RulePrerequisite } from "./single/RulePrerequisite.js"
import { SexPrerequisite } from "./single/SexPrerequisite.js"
import { SexualCharacteristicPrerequisite } from "./single/SexualCharacteristicPrerequisite.js"
import { SocialStatusPrerequisite } from "./single/SocialStatusPrerequisite.js"
import { StatePrerequisite } from "./single/StatePrerequisite.js"
import { TextPrerequisite } from "./single/TextPrerequisite.js"
import {
  BlessedTraditionPrerequisite,
  MagicalTraditionPrerequisite,
} from "./single/TraditionPrerequisite.js"

export const RulePrerequisiteGroup = DB.Enum(import.meta.url, {
  name: "RulePrerequisiteGroup",
  values: () => ({
    Rule: DB.EnumCase({ type: DB.IncludeIdentifier(RulePrerequisite) }),
  }),
})

export const DerivedCharacteristicPrerequisiteGroup = DB.Enum(import.meta.url, {
  name: "DerivedCharacteristicPrerequisiteGroup",
  values: () => ({
    Rule: DB.EnumCase({ type: DB.IncludeIdentifier(RulePrerequisite) }),
  }),
})

export const PublicationPrerequisiteGroup = DB.Enum(import.meta.url, {
  name: "PublicationPrerequisiteGroup",
  values: () => ({
    Publication: DB.EnumCase({ type: DB.IncludeIdentifier(PublicationPrerequisite) }),
  }),
})

export const GeneralPrerequisiteGroup = DB.Enum(import.meta.url, {
  name: "GeneralPrerequisiteGroup",
  values: () => ({
    Sex: DB.EnumCase({ type: DB.IncludeIdentifier(SexPrerequisite) }),
    Race: DB.EnumCase({ type: DB.IncludeIdentifier(RacePrerequisite) }),
    Culture: DB.EnumCase({ type: DB.IncludeIdentifier(CulturePrerequisite) }),
    Pact: DB.EnumCase({ type: DB.IncludeIdentifier(PactPrerequisite) }),
    SocialStatus: DB.EnumCase({ type: DB.IncludeIdentifier(SocialStatusPrerequisite) }),
    State: DB.EnumCase({ type: DB.IncludeIdentifier(StatePrerequisite) }),
    Rule: DB.EnumCase({ type: DB.IncludeIdentifier(RulePrerequisite) }),
    PrimaryAttribute: DB.EnumCase({ type: DB.IncludeIdentifier(PrimaryAttributePrerequisite) }),
    Activatable: DB.EnumCase({ type: DB.IncludeIdentifier(ActivatablePrerequisite) }),
    BlessedTradition: DB.EnumCase({ type: DB.IncludeIdentifier(BlessedTraditionPrerequisite) }),
    MagicalTradition: DB.EnumCase({ type: DB.IncludeIdentifier(MagicalTraditionPrerequisite) }),
    Rated: DB.EnumCase({ type: DB.IncludeIdentifier(RatedPrerequisite) }),
    RatedMinimumNumber: DB.EnumCase({ type: DB.IncludeIdentifier(RatedMinimumNumberPrerequisite) }),
    RatedSum: DB.EnumCase({ type: DB.IncludeIdentifier(RatedSumPrerequisite) }),
    Enhancement: DB.EnumCase({ type: DB.IncludeIdentifier(EnhancementPrerequisite) }),
    Text: DB.EnumCase({ type: DB.IncludeIdentifier(TextPrerequisite) }),
    SexualCharacteristic: DB.EnumCase({
      type: DB.IncludeIdentifier(SexualCharacteristicPrerequisite),
    }),
  }),
})

export const ProfessionPrerequisiteGroup = DB.Enum(import.meta.url, {
  name: "ProfessionPrerequisiteGroup",
  values: () => ({
    Sex: DB.EnumCase({ type: DB.IncludeIdentifier(SexPrerequisite) }),
    Race: DB.EnumCase({ type: DB.IncludeIdentifier(RacePrerequisite) }),
    Culture: DB.EnumCase({ type: DB.IncludeIdentifier(CulturePrerequisite) }),
    Activatable: DB.EnumCase({ type: DB.IncludeIdentifier(ActivatablePrerequisite) }),
    Rated: DB.EnumCase({ type: DB.IncludeIdentifier(RatedPrerequisite) }),
  }),
})

export const AdvantageDisadvantagePrerequisiteGroup = DB.Enum(import.meta.url, {
  name: "AdvantageDisadvantagePrerequisiteGroup",
  values: () => {
    return {
      CommonSuggestedByRCP: DB.EnumCase({ type: null }),
      Sex: DB.EnumCase({ type: DB.IncludeIdentifier(SexPrerequisite) }),
      Race: DB.EnumCase({ type: DB.IncludeIdentifier(RacePrerequisite) }),
      Culture: DB.EnumCase({ type: DB.IncludeIdentifier(CulturePrerequisite) }),
      Pact: DB.EnumCase({ type: DB.IncludeIdentifier(PactPrerequisite) }),
      SocialStatus: DB.EnumCase({ type: DB.IncludeIdentifier(SocialStatusPrerequisite) }),
      State: DB.EnumCase({ type: DB.IncludeIdentifier(StatePrerequisite) }),
      Rule: DB.EnumCase({ type: DB.IncludeIdentifier(RulePrerequisite) }),
      PrimaryAttribute: DB.EnumCase({ type: DB.IncludeIdentifier(PrimaryAttributePrerequisite) }),
      Activatable: DB.EnumCase({ type: DB.IncludeIdentifier(ActivatablePrerequisite) }),
      BlessedTradition: DB.EnumCase({ type: DB.IncludeIdentifier(BlessedTraditionPrerequisite) }),
      MagicalTradition: DB.EnumCase({ type: DB.IncludeIdentifier(MagicalTraditionPrerequisite) }),
      Rated: DB.EnumCase({ type: DB.IncludeIdentifier(RatedPrerequisite) }),
      RatedMinimumNumber: DB.EnumCase({
        type: DB.IncludeIdentifier(RatedMinimumNumberPrerequisite),
      }),
      RatedSum: DB.EnumCase({ type: DB.IncludeIdentifier(RatedSumPrerequisite) }),
      Enhancement: DB.EnumCase({ type: DB.IncludeIdentifier(EnhancementPrerequisite) }),
      Text: DB.EnumCase({ type: DB.IncludeIdentifier(TextPrerequisite) }),
      NoOtherAncestorBloodAdvantage: DB.EnumCase({
        type: AncestorBloodPrerequisite,
      }),
      SexualCharacteristic: DB.EnumCase({
        type: DB.IncludeIdentifier(SexualCharacteristicPrerequisite),
      }),
    }
  },
})

export const ArcaneTraditionPrerequisiteGroup = DB.Enum(import.meta.url, {
  name: "ArcaneTraditionPrerequisiteGroup",
  values: () => ({
    Sex: DB.EnumCase({ type: DB.IncludeIdentifier(SexPrerequisite) }),
    Culture: DB.EnumCase({ type: DB.IncludeIdentifier(CulturePrerequisite) }),
  }),
})

export const PersonalityTraitPrerequisiteGroup = DB.Enum(import.meta.url, {
  name: "PersonalityTraitPrerequisiteGroup",
  values: () => ({
    Race: DB.EnumCase({ type: DB.IncludeIdentifier(RacePrerequisite) }),
    Culture: DB.EnumCase({ type: DB.IncludeIdentifier(CulturePrerequisite) }),
    PersonalityTrait: DB.EnumCase({ type: DB.IncludeIdentifier(PersonalityTraitPrerequisite) }),
    Text: DB.EnumCase({ type: DB.IncludeIdentifier(TextPrerequisite) }),
  }),
})

export const SpellworkPrerequisiteGroup = DB.Enum(import.meta.url, {
  name: "SpellworkPrerequisiteGroup",
  values: () => ({
    Rule: DB.EnumCase({ type: DB.IncludeIdentifier(RulePrerequisite) }),
    Rated: DB.EnumCase({ type: DB.IncludeIdentifier(RatedPrerequisite) }),
  }),
})

export const LiturgyPrerequisiteGroup = DB.Enum(import.meta.url, {
  name: "LiturgyPrerequisiteGroup",
  values: () => ({
    Rule: DB.EnumCase({ type: DB.IncludeIdentifier(RulePrerequisite) }),
  }),
})

export const InfluencePrerequisiteGroup = DB.Enum(import.meta.url, {
  name: "InfluencePrerequisiteGroup",
  values: () => ({
    Influence: DB.EnumCase({ type: DB.IncludeIdentifier(InfluencePrerequisite) }),
    Text: DB.EnumCase({ type: DB.IncludeIdentifier(TextPrerequisite) }),
  }),
})

export const LanguagePrerequisiteGroup = DB.Enum(import.meta.url, {
  name: "LanguagePrerequisiteGroup",
  values: () => ({
    Race: DB.EnumCase({ type: DB.IncludeIdentifier(RacePrerequisite) }),
    Activatable: DB.EnumCase({ type: DB.IncludeIdentifier(ActivatablePrerequisite) }),
    Text: DB.EnumCase({ type: DB.IncludeIdentifier(TextPrerequisite) }),
  }),
})

export const AnimistPowerPrerequisiteGroup = DB.Enum(import.meta.url, {
  name: "AnimistPowerPrerequisiteGroup",
  values: () => ({
    AnimistPower: DB.EnumCase({ type: DB.IncludeIdentifier(AnimistPowerPrerequisite) }),
  }),
})

export const TribePrerequisiteGroup = DB.Enum(import.meta.url, {
  name: "TribePrerequisiteGroup",
  values: () => ({
    Race: DB.EnumCase({ type: DB.IncludeIdentifier(RacePrerequisite) }),
    Culture: DB.EnumCase({ type: DB.IncludeIdentifier(CulturePrerequisite) }),
  }),
})

export const GeodeRitualPrerequisiteGroup = DB.Enum(import.meta.url, {
  name: "GeodeRitualPrerequisiteGroup",
  values: () => ({
    Influence: DB.EnumCase({ type: DB.IncludeIdentifier(InfluencePrerequisite) }),
  }),
})

export const EnhancementPrerequisiteGroup = DB.Enum(import.meta.url, {
  name: "EnhancementPrerequisiteGroup",
  values: () => ({
    Enhancement: DB.EnumCase({ type: DB.IncludeIdentifier(EnhancementPrerequisite) }),
    Rated: DB.EnumCase({ type: DB.IncludeIdentifier(RatedPrerequisite) }),
  }),
})

export const PreconditionGroup = DB.Enum(import.meta.url, {
  name: "PreconditionGroup",
  values: () => ({
    Publication: DB.EnumCase({ type: DB.IncludeIdentifier(PublicationPrerequisite) }),
    SexualCharacteristic: DB.EnumCase({
      type: DB.IncludeIdentifier(SexualCharacteristicPrerequisite),
    }),
  }),
})

File: Sources/OptolithDatabaseSchema/GeneratedTypes/prerequisites/PrerequisiteGroups.swift
//
//  PrerequisiteGroups.swift
//  OptolithDatabaseSchema
//

public enum DerivedCharacteristicPrerequisiteGroup: EntitySubtype {
    case rule(RulePrerequisite)

    private enum CodingKeys: String, CodingKey {
        case tag = "tag"
        case rule = "rule"
    }

    private enum Discriminator: String, Decodable {
        case rule = "Rule"
    }

    public init(from decoder: any Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let tag = try container.decode(Discriminator.self, forKey: .tag)
        switch tag {
        case .rule:
            self = .rule(try container.decode(RulePrerequisite.self, forKey: .rule))
        }
    }
}

public enum PublicationPrerequisiteGroup: EntitySubtype {
    case publication(PublicationPrerequisite)

    private enum CodingKeys: String, CodingKey {
        case tag = "tag"
        case publication = "publication"
    }

    private enum Discriminator: String, Decodable {
        case publication = "Publication"
    }

    public init(from decoder: any Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let tag = try container.decode(Discriminator.self, forKey: .tag)
        switch tag {
        case .publication:
            self = .publication(try container.decode(PublicationPrerequisite.self, forKey: .publication))
        }
    }
}

public enum GeneralPrerequisiteGroup: EntitySubtype {
    case sex(SexPrerequisite)
    case race(RacePrerequisite)
    case culture(CulturePrerequisite)
    case pact(PactPrerequisite)
    case socialStatus(SocialStatusPrerequisite)
    case state(StatePrerequisite)
    case rule(RulePrerequisite)
    case primaryAttribute(PrimaryAttributePrerequisite)
    case activatable(ActivatablePrerequisite)
    case blessedTradition(BlessedTraditionPrerequisite)
    case magicalTradition(MagicalTraditionPrerequisite)
    case rated(RatedPrerequisite)
    case ratedMinimumNumber(RatedMinimumNumberPrerequisite)
    case ratedSum(RatedSumPrerequisite)
    case externalEnhancement(ExternalEnhancementPrerequisite)
    case text(TextPrerequisite)
    case sexualCharacteristic(SexualCharacteristicPrerequisite)

    private enum CodingKeys: String, CodingKey {
        case tag = "tag"
        case sex = "sex"
        case race = "race"
        case culture = "culture"
        case pact = "pact"
        case socialStatus = "social_status"
        case state = "state"
        case rule = "rule"
        case primaryAttribute = "primary_attribute"
        case activatable = "activatable"
        case blessedTradition = "blessed_tradition"
        case magicalTradition = "magical_tradition"
        case rated = "rated"
        case ratedMinimumNumber = "rated_minimum_number"
        case ratedSum = "rated_sum"
        case externalEnhancement = "external_enhancement"
        case text = "text"
        case sexualCharacteristic = "sexual_characteristic"
    }

    private enum Discriminator: String, Decodable {
        case sex = "Sex"
        case race = "Race"
        case culture = "Culture"
        case pact = "Pact"
        case socialStatus = "SocialStatus"
        case state = "State"
        case rule = "Rule"
        case primaryAttribute = "PrimaryAttribute"
        case activatable = "Activatable"
        case blessedTradition = "BlessedTradition"
        case magicalTradition = "MagicalTradition"
        case rated = "Rated"
        case ratedMinimumNumber = "RatedMinimumNumber"
        case ratedSum = "RatedSum"
        case externalEnhancement = "ExternalEnhancement"
        case text = "Text"
        case sexualCharacteristic = "SexualCharacteristic"
    }

    public init(from decoder: any Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let tag = try container.decode(Discriminator.self, forKey: .tag)
        switch tag {
        case .sex:
            self = .sex(try container.decode(SexPrerequisite.self, forKey: .sex))
        case .race:
            self = .race(try container.decode(RacePrerequisite.self, forKey: .race))
        case .culture:
            self = .culture(try container.decode(CulturePrerequisite.self, forKey: .culture))
        case .pact:
            self = .pact(try container.decode(PactPrerequisite.self, forKey: .pact))
        case .socialStatus:
            self = .socialStatus(try container.decode(SocialStatusPrerequisite.self, forKey: .socialStatus))
        case .state:
            self = .state(try container.decode(StatePrerequisite.self, forKey: .state))
        case .rule:
            self = .rule(try container.decode(RulePrerequisite.self, forKey: .rule))
        case .primaryAttribute:
            self = .primaryAttribute(try container.decode(PrimaryAttributePrerequisite.self, forKey: .primaryAttribute))
        case .activatable:
            self = .activatable(try container.decode(ActivatablePrerequisite.self, forKey: .activatable))
        case .blessedTradition:
            self = .blessedTradition(try container.decode(BlessedTraditionPrerequisite.self, forKey: .blessedTradition))
        case .magicalTradition:
            self = .magicalTradition(try container.decode(MagicalTraditionPrerequisite.self, forKey: .magicalTradition))
        case .rated:
            self = .rated(try container.decode(RatedPrerequisite.self, forKey: .rated))
        case .ratedMinimumNumber:
            self = .ratedMinimumNumber(try container.decode(RatedMinimumNumberPrerequisite.self, forKey: .ratedMinimumNumber))
        case .ratedSum:
            self = .ratedSum(try container.decode(RatedSumPrerequisite.self, forKey: .ratedSum))
        case .externalEnhancement:
            self = .externalEnhancement(try container.decode(ExternalEnhancementPrerequisite.self, forKey: .externalEnhancement))
        case .text:
            self = .text(try container.decode(TextPrerequisite.self, forKey: .text))
        case .sexualCharacteristic:
            self = .sexualCharacteristic(try container.decode(SexualCharacteristicPrerequisite.self, forKey: .sexualCharacteristic))
        }
    }
}

public enum ProfessionPrerequisiteGroup: EntitySubtype {
    case sex(SexPrerequisite)
    case race(RacePrerequisite)
    case culture(CulturePrerequisite)
    case activatable(ActivatablePrerequisite)
    case rated(RatedPrerequisite)

    private enum CodingKeys: String, CodingKey {
        case tag = "tag"
        case sex = "sex"
        case race = "race"
        case culture = "culture"
        case activatable = "activatable"
        case rated = "rated"
    }

    private enum Discriminator: String, Decodable {
        case sex = "Sex"
        case race = "Race"
        case culture = "Culture"
        case activatable = "Activatable"
        case rated = "Rated"
    }

    public init(from decoder: any Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let tag = try container.decode(Discriminator.self, forKey: .tag)
        switch tag {
        case .sex:
            self = .sex(try container.decode(SexPrerequisite.self, forKey: .sex))
        case .race:
            self = .race(try container.decode(RacePrerequisite.self, forKey: .race))
        case .culture:
            self = .culture(try container.decode(CulturePrerequisite.self, forKey: .culture))
        case .activatable:
            self = .activatable(try container.decode(ActivatablePrerequisite.self, forKey: .activatable))
        case .rated:
            self = .rated(try container.decode(RatedPrerequisite.self, forKey: .rated))
        }
    }
}

public enum AdvantageDisadvantagePrerequisiteGroup: EntitySubtype {
    case commonSuggestedByRCP
    case sex(SexPrerequisite)
    case race(RacePrerequisite)
    case culture(CulturePrerequisite)
    case pact(PactPrerequisite)
    case socialStatus(SocialStatusPrerequisite)
    case state(StatePrerequisite)
    case rule(RulePrerequisite)
    case primaryAttribute(PrimaryAttributePrerequisite)
    case activatable(ActivatablePrerequisite)
    case blessedTradition(BlessedTraditionPrerequisite)
    case magicalTradition(MagicalTraditionPrerequisite)
    case rated(RatedPrerequisite)
    case ratedMinimumNumber(RatedMinimumNumberPrerequisite)
    case ratedSum(RatedSumPrerequisite)
    case externalEnhancement(ExternalEnhancementPrerequisite)
    case text(TextPrerequisite)
    case noOtherAncestorBloodAdvantage(AncestorBloodPrerequisite)
    case sexualCharacteristic(SexualCharacteristicPrerequisite)

    private enum CodingKeys: String, CodingKey {
        case tag = "tag"
        case commonSuggestedByRCP = "common_suggested_by_rcp"
        case sex = "sex"
        case race = "race"
        case culture = "culture"
        case pact = "pact"
        case socialStatus = "social_status"
        case state = "state"
        case rule = "rule"
        case primaryAttribute = "primary_attribute"
        case activatable = "activatable"
        case blessedTradition = "blessed_tradition"
        case magicalTradition = "magical_tradition"
        case rated = "rated"
        case ratedMinimumNumber = "rated_minimum_number"
        case ratedSum = "rated_sum"
        case externalEnhancement = "external_enhancement"
        case text = "text"
        case noOtherAncestorBloodAdvantage = "no_other_ancestor_blood_advantage"
        case sexualCharacteristic = "sexual_characteristic"
    }

    private enum Discriminator: String, Decodable {
        case commonSuggestedByRCP = "CommonSuggestedByRCP"
        case sex = "Sex"
        case race = "Race"
        case culture = "Culture"
        case pact = "Pact"
        case socialStatus = "SocialStatus"
        case state = "State"
        case rule = "Rule"
        case primaryAttribute = "PrimaryAttribute"
        case activatable = "Activatable"
        case blessedTradition = "BlessedTradition"
        case magicalTradition = "MagicalTradition"
        case rated = "Rated"
        case ratedMinimumNumber = "RatedMinimumNumber"
        case ratedSum = "RatedSum"
        case externalEnhancement = "ExternalEnhancement"
        case text = "Text"
        case noOtherAncestorBloodAdvantage = "NoOtherAncestorBloodAdvantage"
        case sexualCharacteristic = "SexualCharacteristic"
    }

    public init(from decoder: any Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let tag = try container.decode(Discriminator.self, forKey: .tag)
        switch tag {
        case .commonSuggestedByRCP:
            self = .commonSuggestedByRCP
        case .sex:
            self = .sex(try container.decode(SexPrerequisite.self, forKey: .sex))
        case .race:
            self = .race(try container.decode(RacePrerequisite.self, forKey: .race))
        case .culture:
            self = .culture(try container.decode(CulturePrerequisite.self, forKey: .culture))
        case .pact:
            self = .pact(try container.decode(PactPrerequisite.self, forKey: .pact))
        case .socialStatus:
            self = .socialStatus(try container.decode(SocialStatusPrerequisite.self, forKey: .socialStatus))
        case .state:
            self = .state(try container.decode(StatePrerequisite.self, forKey: .state))
        case .rule:
            self = .rule(try container.decode(RulePrerequisite.self, forKey: .rule))
        case .primaryAttribute:
            self = .primaryAttribute(try container.decode(PrimaryAttributePrerequisite.self, forKey: .primaryAttribute))
        case .activatable:
            self = .activatable(try container.decode(ActivatablePrerequisite.self, forKey: .activatable))
        case .blessedTradition:
            self = .blessedTradition(try container.decode(BlessedTraditionPrerequisite.self, forKey: .blessedTradition))
        case .magicalTradition:
            self = .magicalTradition(try container.decode(MagicalTraditionPrerequisite.self, forKey: .magicalTradition))
        case .rated:
            self = .rated(try container.decode(RatedPrerequisite.self, forKey: .rated))
        case .ratedMinimumNumber:
            self = .ratedMinimumNumber(try container.decode(RatedMinimumNumberPrerequisite.self, forKey: .ratedMinimumNumber))
        case .ratedSum:
            self = .ratedSum(try container.decode(RatedSumPrerequisite.self, forKey: .ratedSum))
        case .externalEnhancement:
            self = .externalEnhancement(try container.decode(ExternalEnhancementPrerequisite.self, forKey: .externalEnhancement))
        case .text:
            self = .text(try container.decode(TextPrerequisite.self, forKey: .text))
        case .noOtherAncestorBloodAdvantage:
            self = .noOtherAncestorBloodAdvantage(try container.decode(AncestorBloodPrerequisite.self, forKey: .noOtherAncestorBloodAdvantage))
        case .sexualCharacteristic:
            self = .sexualCharacteristic(try container.decode(SexualCharacteristicPrerequisite.self, forKey: .sexualCharacteristic))
        }
    }
}

public enum ArcaneTraditionPrerequisiteGroup: EntitySubtype {
    case sex(SexPrerequisite)
    case culture(CulturePrerequisite)

    private enum CodingKeys: String, CodingKey {
        case tag = "tag"
        case sex = "sex"
        case culture = "culture"
    }

    private enum Discriminator: String, Decodable {
        case sex = "Sex"
        case culture = "Culture"
    }

    public init(from decoder: any Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let tag = try container.decode(Discriminator.self, forKey: .tag)
        switch tag {
        case .sex:
            self = .sex(try container.decode(SexPrerequisite.self, forKey: .sex))
        case .culture:
            self = .culture(try container.decode(CulturePrerequisite.self, forKey: .culture))
        }
    }
}

public enum PersonalityTraitPrerequisiteGroup: EntitySubtype {
    case culture(CulturePrerequisite)
    case text(TextPrerequisite)

    private enum CodingKeys: String, CodingKey {
        case tag = "tag"
        case culture = "culture"
        case text = "text"
    }

    private enum Discriminator: String, Decodable {
        case culture = "Culture"
        case text = "Text"
    }

    public init(from decoder: any Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let tag = try container.decode(Discriminator.self, forKey: .tag)
        switch tag {
        case .culture:
            self = .culture(try container.decode(CulturePrerequisite.self, forKey: .culture))
        case .text:
            self = .text(try container.decode(TextPrerequisite.self, forKey: .text))
        }
    }
}

public enum SpellworkPrerequisiteGroup: EntitySubtype {
    case rule(RulePrerequisite)
    case rated(RatedPrerequisite)

    private enum CodingKeys: String, CodingKey {
        case tag = "tag"
        case rule = "rule"
        case rated = "rated"
    }

    private enum Discriminator: String, Decodable {
        case rule = "Rule"
        case rated = "Rated"
    }

    public init(from decoder: any Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let tag = try container.decode(Discriminator.self, forKey: .tag)
        switch tag {
        case .rule:
            self = .rule(try container.decode(RulePrerequisite.self, forKey: .rule))
        case .rated:
            self = .rated(try container.decode(RatedPrerequisite.self, forKey: .rated))
        }
    }
}

public enum LiturgyPrerequisiteGroup: EntitySubtype {
    case rule(RulePrerequisite)

    private enum CodingKeys: String, CodingKey {
        case tag = "tag"
        case rule = "rule"
    }

    private enum Discriminator: String, Decodable {
        case rule = "Rule"
    }

    public init(from decoder: any Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let tag = try container.decode(Discriminator.self, forKey: .tag)
        switch tag {
        case .rule:
            self = .rule(try container.decode(RulePrerequisite.self, forKey: .rule))
        }
    }
}

public enum InfluencePrerequisiteGroup: EntitySubtype {
    case influence(InfluencePrerequisite)
    case text(TextPrerequisite)

    private enum CodingKeys: String, CodingKey {
        case tag = "tag"
        case influence = "influence"
        case text = "text"
    }

    private enum Discriminator: String, Decodable {
        case influence = "Influence"
        case text = "Text"
    }

    public init(from decoder: any Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let tag = try container.decode(Discriminator.self, forKey: .tag)
        switch tag {
        case .influence:
            self = .influence(try container.decode(InfluencePrerequisite.self, forKey: .influence))
        case .text:
            self = .text(try container.decode(TextPrerequisite.self, forKey: .text))
        }
    }
}

public enum LanguagePrerequisiteGroup: EntitySubtype {
    case race(RacePrerequisite)
    case activatable(ActivatablePrerequisite)
    case text(TextPrerequisite)

    private enum CodingKeys: String, CodingKey {
        case tag = "tag"
        case race = "race"
        case activatable = "activatable"
        case text = "text"
    }

    private enum Discriminator: String, Decodable {
        case race = "Race"
        case activatable = "Activatable"
        case text = "Text"
    }

    public init(from decoder: any Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let tag = try container.decode(Discriminator.self, forKey: .tag)
        switch tag {
        case .race:
            self = .race(try container.decode(RacePrerequisite.self, forKey: .race))
        case .activatable:
            self = .activatable(try container.decode(ActivatablePrerequisite.self, forKey: .activatable))
        case .text:
            self = .text(try container.decode(TextPrerequisite.self, forKey: .text))
        }
    }
}

public enum AnimistPowerPrerequisiteGroup: EntitySubtype {
    case animistPower(AnimistPowerPrerequisite)

    private enum CodingKeys: String, CodingKey {
        case tag = "tag"
        case animistPower = "animist_power"
    }

    private enum Discriminator: String, Decodable {
        case animistPower = "AnimistPower"
    }

    public init(from decoder: any Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let tag = try container.decode(Discriminator.self, forKey: .tag)
        switch tag {
        case .animistPower:
            self = .animistPower(try container.decode(AnimistPowerPrerequisite.self, forKey: .animistPower))
        }
    }
}

public enum GeodeRitualPrerequisiteGroup: EntitySubtype {
    case influence(InfluencePrerequisite)

    private enum CodingKeys: String, CodingKey {
        case tag = "tag"
        case influence = "influence"
    }

    private enum Discriminator: String, Decodable {
        case influence = "Influence"
    }

    public init(from decoder: any Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let tag = try container.decode(Discriminator.self, forKey: .tag)
        switch tag {
        case .influence:
            self = .influence(try container.decode(InfluencePrerequisite.self, forKey: .influence))
        }
    }
}

public enum EnhancementPrerequisiteGroup: EntitySubtype {
    case internalEnhancement(InternalEnhancementPrerequisite)

    private enum CodingKeys: String, CodingKey {
        case tag = "tag"
        case internalEnhancement = "internal_enhancement"
    }

    private enum Discriminator: String, Decodable {
        case internalEnhancement = "InternalEnhancement"
    }

    public init(from decoder: any Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let tag = try container.decode(Discriminator.self, forKey: .tag)
        switch tag {
        case .internalEnhancement:
            self = .internalEnhancement(try container.decode(InternalEnhancementPrerequisite.self, forKey: .internalEnhancement))
        }
    }
}

public enum PreconditionGroup: EntitySubtype {
    case publication(PublicationPrerequisite)
    case sexualCharacteristic(SexualCharacteristicPrerequisite)

    private enum CodingKeys: String, CodingKey {
        case tag = "tag"
        case publication = "publication"
        case sexualCharacteristic = "sexual_characteristic"
    }

    private enum Discriminator: String, Decodable {
        case publication = "Publication"
        case sexualCharacteristic = "SexualCharacteristic"
    }

    public init(from decoder: any Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let tag = try container.decode(Discriminator.self, forKey: .tag)
        switch tag {
        case .publication:
            self = .publication(try container.decode(PublicationPrerequisite.self, forKey: .publication))
        case .sexualCharacteristic:
            self = .sexualCharacteristic(try container.decode(SexualCharacteristicPrerequisite.self, forKey: .sexualCharacteristic))
        }
    }
}
